Add support for querying the events-log plugin

opalmer opened the pull request:
This PR adds support for querying information about Gerrit events via the optional events-log plugin.  The structures this PR adds could also be used when parsing output from the built-in `gerrit stream-events` command over ssh.

opalmer commented:
@andygrunwald PTAL.  It was not clear to me if there was anything for or against providing a function to parse the output of an optional plugin so I thought I'd open up a PR to discuss it.  Even without `GetEvents()` it would seem that providing the structures of a Gerrit event would be pretty useful since they could be used for parsing the output of the stream-events command.

Also, how do you think this should be tested?  Most of the tests today talk to an external server which for this code won't work since it requires auth.  Have you thought about having the tests spin up an http server that can provide responses to requests?

andygrunwald commented:
Thank you @opalmer.
Is this PR for the support of [this event-log plugin](https://gerrit.googlesource.com/plugins/events-log/)?

I don`t have a problem with adding support for plugins. The plugins needs to "kind of" popular, but as far as i can see / know this applies to event-log.
And as you said  the structs are able to reuse. So this is fine :)

Related to testing:
Yes, you can bootstrap an own http server for tests. I did this in a different project. For this is have a [setup() and tearDown()](https://github.com/andygrunwald/go-jira/blob/master/jira_test.go#L36) func. Those will bootstrap the http server. And here is a test how to use this: See [TestProjectService_GetList](https://github.com/andygrunwald/go-jira/blob/master/project_test.go#L10). If you have a look at different tests in the [go-jira](https://github.com/andygrunwald/go-jira) you will encounter a pattern there.
This can be applied to this client library as well.
Do you plan to write tests for this PR / would you be so nice to write a few unit tests?

And it would be cool if you can add a note into the README (e.g. Features section) that we have support for this plugin as well.

opalmer commented:
> Is this PR for the support of this event-log plugin?

That's the one.

> I don`t have a problem with adding support for plugins. The plugins needs to "kind of" popular, but as far as i can see / know this applies to event-log. And as you said the structs are able to reuse. So this is fine :)

Awesome, thanks!  I'm happy to add some notes to the README alone those lines if you think that would be appropriate for future contributors.

> Do you plan to write tests for this PR / would you be so nice to write a few unit tests?

Sure happy to though it may be a bit before I can get it done (have some other things I'm working on at the moment).

> And it would be cool if you can add a note into the README (e.g. Features section) that we have support for this plugin as well.

Will do.

opalmer commented:
@andygrunwald PTAL.  Added tests, updated the README and fixed a couple of bugs.  

On a side note, I'll probably be opening another PR shortly to update the imports in the tests from `github.com/andygrunwald/go-gerrit` to `.`.  Without this I had to do some $GOPATH manipulation in order for the tests to use my forked repository.

andygrunwald commented:
Thank you.  I merged all changes.

Related to the tests:
I am fine with renaming the test package to `gerrit_test`.
This would have the big benefit that we are only able to test the client api and no internal state (this would bring us the behaviour to create faulty tests).
Go for it :)

opalmer commented:
The current name of the test package is gerrit_test though and I was
proposing that be changed to gerrit.  If you prefer to keep the existing
name, which I don't object to for the reason you mentioned, are you ok with
an update to the imports so they're relative instead (`.`)?

On Aug 2, 2016 2:31 AM, "Andy Grunwald" <notifications@github.com> wrote:

> Thank you. I merged all changes.
>
> Related to the tests:
> I am fine with renaming the test package to gerrit_test.
> This would have the big benefit that we are only able to test the client
> api and no internal state (this would bring us the behaviour to create
> faulty tests).
> Go for it :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andygrunwald/go-gerrit/pull/10#issuecomment-236812813>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPkjaXQJ7rWJwJSH0ioK6urk6S0NTedks5qbuQ_gaJpZM4JY1xK>
> .
>

andygrunwald commented:
@opalmer I don`t like the change with the import path. I understand your problem related to the fork. Would it be an option for you to 
a) apply the `git remotes` way like in the blog post you mentioned
or (maybe better)
b) be a contributor with write access to this project? Then i would add you to this.

opalmer commented:
@andygrunwald sorry for the delay, I've been traveling.  Totally understand wanting to keep the import path as is so no argument there.

Options B would be nice if you're ok with it 😄 .  I'll still open PRs of course but having write access would help with development.  Another added benefit is if you're ever away I can sub in for bug fixes and code reviews. On a related side note, it's probably a good idea if I try out option A anyway and add it to the README once I give it a try.  The setup is not all that complicated but a little documentation around the procedure, and why it's necessary, could help a future contributor.

andygrunwald commented:
No worry.
I`ve added you as an collaborator. You just need to accept the invite. Welcome :)
It would be good to open PRs for changes that we or other people are able to review them. Sometimes magic things can happen :)

Looking forward to your changes.

shurcooL commented on gerrit.go:
```diff
@@ -265,18 +271,22 @@ func (c *Client) DeleteRequest(urlStr string, body interface{}) (*Response, erro
+			// +1 to catch the \n as well
+			body = body[(index + 1):]
+		}
 	}
```
This code can be significantly simplified without changing behavior.

If body has `")]}'\n"` prefix, there's no need to use `bytes.IndexByte` to find the first `'\n'` character, it's known to be at position 3 because `bytes.HasPrefix(body, []byte(")]}'\n"))` was true.

Also, since this is called often, it's probably a good idea to factor out `[]byte(")]}'\n")` into a single package scope variable, instead of potentially allocating once per `RemoveMagicPrefixLine` call.

```Go
func RemoveMagicPrefixLine(body []byte) []byte {
	if bytes.HasPrefix(body, magicPrefix) {
		return body[5:]
	}
	return body
}

var magicPrefix = []byte(")]}'\n")
```

opalmer replied:
@shurcooL, no issues here if you want to go ahead and make the changes. I'd do it myself but I don't currently have my local host setup to make the change and test it.

shurcooL replied:
@opalmer I already made a PR that applies this suggestion, see #36.

opalmer replied:
Ah my bad sorry about that. Reviewed, thanks!

andygrunwald replied:
Thanks both of you @opalmer and @shurcooL
